gson/src/test/java/com/google/gson/JsonObjectTest.java

```diff
@@ -200,5 +210,92 @@ public void testKeySet() {
     assertEquals(2, a.keySet().size());
     assertTrue(a.keySet().contains("foo"));
     assertTrue(a.keySet().contains("bar"));
+
+    a.addProperty("1", true);
+    a.addProperty("2", false);
+
+    // Insertion order should be preserved by keySet()
+    Deque<String> expectedKeys = new ArrayDeque<>(Arrays.asList("foo", "bar", "1", "2"));
+    assertEquals(new ArrayList<>(expectedKeys), new ArrayList<>(a.keySet()));
+    Iterator<String> iterator = a.keySet().iterator();
+
+    // Remove keys one by one
+    for (int i = a.size(); i >= 1; i--) {
+      assertTrue(iterator.hasNext());
+      assertEquals(expectedKeys.getFirst(), iterator.next());
+      iterator.remove();
+      expectedKeys.removeFirst();
+
+      assertEquals(i - 1, a.size());
+      // Note: Must wrap in ArrayList because Deque implementations do not implement `equals`
```

Comment:
Maybe move this comment before line 219, since that's the first time you compare `expectedKeys`.